[docs] program examples (#186)
solana-foundation/developer-content: docs/programs/examples.md

---
date: Apr 26, 2024
title: "Program Examples"
description:
  "A list of Solana program examples in different languages and frameworks,
  which can help you learn and use as reference for your own projects."
tags:
  - quickstart
  - program
  - anchor
  - javascript
  - native
  - rust
  - solang
  - token22
  - token extensions
keywords:
  - rust
  - cargo
  - toml
  - program
  - tutorial
  - intro to solana development
  - blockchain developer
  - blockchain tutorial
  - web3 developer
  - anchor
  - solang
---

> The "Solana Program Examples" repository on GitHub offers several subfolders,
> each containing code examples for different Solana programming paradigms and
> languages, designed to help developers learn and experiment with Solana
> blockchain development.

Reviewer: Let's put the link here, and then the repo path where the link was:

```suggestion
The
"[Solana Program Examples](https://github.com/solana-developers/program-examples)"
repository on GitHub offers several subfolders, each containing code examples
for different Solana programming paradigms and languages, designed to help
developers learn and experiment with Solana blockchain development.
```

> You can find the examples in the
> [Solana Program Examples repository](https://github.com/solana-developers/program-examples)
> together with README files that explain you how to run the different examples.

Reviewer: ```suggestion
You can find the examples in the `solana-developers/program-examples` together
with README files that explain you how to run the different examples. Most
```

> Most examples are self-contained and are available in Native rust,
> [Anchor](https://www.anchor-lang.com/docs/installation),
> [Seahorse](https://seahorse-lang.org/) and
> [Solang](https://solana.com/de/developers/guides/solang/getting-started).

Reviewer: let's not put solang in here...

```suggestion
Most examples are self-contained and are available in Native rust,
[Anchor](https://www.anchor-lang.com/docs/installation) (rust), and
[Seahorse](https://seahorse-lang.org/) (python).
```

> It also contains a list of examples that we would love to see as contributions.

Reviewer: let's link directly to it!

```suggestion
It also contains a list of examples that we would love to
[see as contributions](https://github.com/solana-developers/program-examples?tab=readme-ov-file#examples-wed-love-to-see).
```

> Here’s a brief description of what each subfolder contains:

Reviewer: ```suggestion
Within the repo you will find the following subfolder, each with assorted
example programs within them:

- [`basics`](#basics)
- [`compression`](#compression)
- [`oracles`](#oracles)
- [`tokens`](#tokens)
```

| **[Program Derived Addresses](https://github.com/solana-developers/program-examples/tree/main/basics/program-derived-addresses)** | Shows how to use seeds to refer to a PDA and save data in it.                                    | Native, Anchor, Solang                      |
| **[Realloc](https://github.com/solana-developers/program-examples/tree/main/basics/realloc)**                                     | Shows you how to increase and decrease the size of an existing account.                          | Native, Anchor                              |
| **[Rent](https://github.com/solana-developers/program-examples/tree/main/basics/rent)**                                           | Here you will learn how to calculate rent requirements within a program.                         | Native, Anchor, Solang                      |
| **[Repository Layout](https://github.com/solana-developers/program-examples/tree/main/basics/repository-layout)**                 | Recommendations on how to structure your program layout.                                         | Native, Anchor                              |

> | **[Transfer Sol](https://github.com/solana-developers/program-examples/tree/main/basics/transfer-sol)**                           | Different methods of transfering sol for system accounts and PDAs.                               | Native, Anchor, Solang, Seahorse            |

Reviewer: ```suggestion
| **[Transfer SOL](https://github.com/solana-developers/program-examples/tree/main/basics/transfer-sol)**                           | Different methods of transferring SOL for system accounts and PDAs.                               | Native, Anchor, Solang, Seahorse            |
```

Contains a series of examples that demonstrate how to use account compression on
Solana. Mainly focused on compressed NFTs (cNFTs).

| Example Name                                                                                                    | Description                                                                         | Language |
| --------------------------------------------------------------------------------------------------------------- | ----------------------------------------------------------------------------------- | -------- |
| **[cNFT-burn](https://github.com/solana-developers/program-examples/tree/main/compression/cnft-burn)**          | To destroy a cNFT it can be burnt. This examples shows how to do that in a program. | Anchor   |
| **[cNFT-solang](https://github.com/solana-developers/program-examples/tree/main/compression/cnft-solang)**      | How to ming a cNFT in Solang                                                        | Solang   |
| **[cNFT-Vault](https://github.com/solana-developers/program-examples/tree/main/compression/cnft-vault/anchor)** | How to custody a cNFT in a program and send it out again.                           | Anchor   |

> | **[cutils](https://github.com/solana-developers/program-examples/tree/main/compression/cutils)**                | A suite utils to for example mint and verfiy cNFTs in a program.                    | Anchor   |

Reviewer: ```suggestion
| **[cutils](https://github.com/solana-developers/program-examples/tree/main/compression/cutils)**                | A suite utils to for example mint and verify cNFTs in a program.                    | Anchor   |
```

| **[Create Token](https://github.com/solana-developers/program-examples/tree/main/tokens/create-token)**             | How to create a token and add metaplex metadata to it.                                           | Anchor, Native, Solang           |
| **[Nft Minter](https://github.com/solana-developers/program-examples/tree/main/tokens/nft-minter)**                 | Minting only one amount of a token and then removing the mint authority.                         | Anchor, Native, Solang           |
| **[PDA Mint Authority](https://github.com/solana-developers/program-examples/tree/main/tokens/pda-mint-authority)** | Shows you how to change the mint authority of a mint, to mint tokens from within a program.      | Anchor, Native, Solang           |
| **[SPL Token Minter](https://github.com/solana-developers/program-examples/tree/main/tokens/spl-token-minter)**     | Explains how to use Associated Token Accounts to be able to keep track of token accounts.        | Anchor, Native, Solang           |

> | **[Token Swap](https://github.com/solana-developers/program-examples/tree/main/tokens/token-swap)**                 | Extensive example that shows you how to build a AMM (automted market maker) pool for SPL tokens. | Anchor                           |

Reviewer: ```suggestion
| **[Token Swap](https://github.com/solana-developers/program-examples/tree/main/tokens/token-swap)**                 | Extensive example that shows you how to build a AMM (automated market maker) pool for SPL tokens. | Anchor                           |
```

---
date: Apr 26, 2024
title: "Program Examples"
description:
  "A list of Solana program examples in different languages and frameworks,
  which can help you learn and use as reference for your own projects."
tags:
  - quickstart
  - program
  - anchor
  - javascript
  - native
  - rust
  - solang
  - token22
  - token extensions
keywords:
  - rust
  - cargo
  - toml
  - program
  - tutorial
  - intro to solana development
  - blockchain developer
  - blockchain tutorial
  - web3 developer
  - anchor
  - solang
---

The
"[Solana Program Examples](https://github.com/solana-developers/program-examples)"
repository on GitHub offers several subfolders, each containing code examples
for different Solana programming paradigms and languages, designed to help
developers learn and experiment with Solana blockchain development.

You can find the examples in the `solana-developers/program-examples` together
with README files that explain you how to run the different examples. Most
examples are self-contained and are available in Native rust,
[Anchor](https://www.anchor-lang.com/docs/installation),

> [Seahorse](https://seahorse-lang.org/) and it also contains a list of examples

Reviewer 2: Just a suggestion that since Seahorse is looking for a permanent maintainer, we hold off to avoid steering people in this direction. We can refocus on Seahorse once we have a maintainer and it's up to date with Solana CLI releases.

Reviewer: Anvit took over being the maintainer a few months ago. Is he not actually maintaining it?

> examples are self-contained and are available in Native rust,
> [Anchor](https://www.anchor-lang.com/docs/installation),
> [Seahorse](https://seahorse-lang.org/) and it also contains a list of examples

Reviewer 2: ```suggestion
examples are self-contained and are available in native Rust (ie, with no framework) and using [Anchor](https://www.anchor-lang.com/docs/installation), and it also contains a list of examples
```
